src/idl/tests/annotation.c

```diff
@@ -67,6 +67,124 @@ CU_Test(idl_annotation, id_member)
   CU_ASSERT_FATAL(idl_is_member(c));
   CU_ASSERT_EQUAL(c->id.annotation, IDL_ID);
   CU_ASSERT_EQUAL(c->id.value, 1);
+
+  idl_delete_pstate(pstate);
+}
+
+CU_Test(idl_annotation, optional)
+{
+  idl_retcode_t ret;
+  idl_pstate_t *pstate = NULL;
+  idl_struct_t *s;
+  idl_member_t *c;
+  idl_declarator_t *d;
+  idl_union_t *u;
+  idl_case_t *uc;
+  const char structstr[] = "struct s {\n"
+                           "  char c_none;\n"
+                           "  @optional(false) char c_false;\n"
+                           "  @optional char c_default;\n"
+                           "  @optional(true) char c_true;\n"
+                           "  @optional char c_1_default, c_2_default;\n"
+                           "};\n";
+
+  ret = parse_string(IDL_FLAG_ANNOTATIONS, structstr, &pstate);
+  CU_ASSERT_EQUAL_FATAL(ret, IDL_RETCODE_OK);
+  CU_ASSERT_PTR_NOT_NULL_FATAL(pstate);
+  assert(pstate);
+  s = (idl_struct_t *)pstate->root;
+  CU_ASSERT_PTR_NOT_NULL_FATAL(s);
+  CU_ASSERT_FATAL(idl_is_struct(s));
+  int n = 0;
+  assert(s);
+  IDL_FOREACH(c, s->members) {
+    CU_ASSERT_PTR_NOT_NULL_FATAL(c);
+    CU_ASSERT_FATAL(idl_is_member(c));
+    CU_ASSERT_EQUAL(idl_is_optional(c), n > 1);
+    IDL_FOREACH(d, c->declarators) {
+      CU_ASSERT_EQUAL(idl_is_optional(d), n > 1);
+    }
+    n++;
+  }
+
+  idl_delete_pstate(pstate);
+
+  const char unionstr[] = "union u switch (long) {\n"
+                          "  case 0: long l;\n"
+                          "  case 2: @optional double d;\n"
+                          "};\n";
+
+  ret = parse_string(IDL_FLAG_ANNOTATIONS, unionstr, &pstate);
+  CU_ASSERT_EQUAL_FATAL(ret, IDL_RETCODE_OK);
+  CU_ASSERT_PTR_NOT_NULL_FATAL(pstate);
+
+  assert(pstate);
+  u = (idl_union_t *)pstate->root;
+  CU_ASSERT_PTR_NOT_NULL_FATAL(u);
+  CU_ASSERT_FATAL(idl_is_union(u));
+
+  n = 0;
+  assert(u);
+  IDL_FOREACH(uc, u->cases) {
+    CU_ASSERT_PTR_NOT_NULL_FATAL(uc);
+    CU_ASSERT_FATAL(idl_is_case(uc));
+    CU_ASSERT_EQUAL(idl_is_optional(uc), n++ > 0);
+  }
+
+  idl_delete_pstate(pstate);
+
+  const char *errstr[] = {"@optional stuct s {\n"
+                          "  char c;\n"
+                          "};\n",
+                          "enum e {\n"
+                          "  e_0,\n"
+                          "  @optional e_1,\n"
+                          "  e_2,\n"
+                          "  e_3\n"
+                          "};\n"};
+
```

Comment:
Maybe add a test for invalid arguments, like `@optional(1)`, `@optional('true')` ?

Reply:
The number of test cases will be expanded a lot in the next iteration.